runtime/nvqir/custatevec/CuStateVecCircuitSimulator.cu

```diff
@@ -324,11 +351,38 @@ protected:
       initializeDeviceStateVector<<<n_blocks, threads_per_block>>>(
           reinterpret_cast<CudaDataType *>(otherState), (1UL << count));
     } else {
-
-      // FIXME Handle case where data is already on GPU
-      HANDLE_CUDA_ERROR(cudaMemcpy(otherState, state,
+      // Check if the pointer is a device pointer
+      cudaPointerAttributes attributes;
+      cudaPointerGetAttributes(&attributes, state);
+
+      if (attributes.type == cudaMemoryTypeDevice) {
+        int currentDevice;
+        HANDLE_CUDA_ERROR(cudaGetDevice(&currentDevice));
+
+        if (attributes.device != currentDevice) {
+          // Memory is on a different GPU
+          // Set the device to the device where the memory is located
+          HANDLE_CUDA_ERROR(cudaSetDevice(attributes.device));
+
+          // Perform device to device copy
+          HANDLE_CUDA_ERROR(cudaMemcpy(otherState, state,
+                                    stateDimension * sizeof(CudaDataType),
+                                    cudaMemcpyDeviceToDevice));
+
+          // Restore the current device
+          HANDLE_CUDA_ERROR(cudaSetDevice(currentDevice));
+        } else {
+          // Memory is on the same GPU
+          HANDLE_CUDA_ERROR(cudaMemcpy(otherState, state,
+                                    stateDimension * sizeof(CudaDataType),
+                                    cudaMemcpyDeviceToDevice));
```

Comment:
I wonder if we need a copy here. `otherState` is only needed for the below kron function where it is passed as a const argument anyway.